### PR TITLE
Adjusting mps writer to the correct structure regarding integer variables declaration

### LINKED ISSUE #1132: mps writer does not mark discrete variables
As described in the MPS documention below, discrete variables should be separated from continuous variable through the use of markers. It does not seem as though the current MPS writer does this.

http://lpsolve.sourceforge.net/5.0/mps-format.htm

### PR DESCRIPTION
**Pull request by Raphael-D-F-Gomes**
## Fixes #1132

This pull request aims to fix the bug described in issue #1132. It involves making simple changes to the MPS writer structure to include the separation of discrete variables following the MPS format (https://lpsolve.sourceforge.net/5.0/mps-format.htm). It wasn't added tests to the bug fix, due to the lack of a mps reader function in the library. 

## Summary/Motivation:
I use the Pyomo library to write mixed-integer linear problems (MILP) to MPS files in my work. I noticed there was a problem in the model, as some binary variables were resulting in floating-point numbers. After researching the Pyomo MPS writer structure, I realized that the MPS file wasn't being exported correctly. So, I expanded the 'ProblemWriter_mps' class and adjusted it to the correct format. These changes worked for me, and I believe they could be a valuable contribution to the library. I created an example using PuLP to read the MPS file exported by Pyomo. 

Consider the MILP problem: 

```
from pyomo.environ import *
from pyomo.opt.base import ProblemFormat

from pulp import LpProblem

model = ConcreteModel("Example-mix-integer-linear-problem")

model.x1 = Var(within=NonNegativeIntegers)  # Integer variable
model.x2 = Var(within=NonNegativeReals)      # Continuous variable

model.obj = Objective(expr=3 * model.x1 + 2 * model.x2, sense=minimize)

model.const1 = Constraint(expr=4 * model.x1 + 3 * model.x2 >= 10)
model.const2 = Constraint(expr=model.x1 + 2 * model.x2 <= 7)

mps_path = 'example_milp.mps'
model.write(mps_path, format=ProblemFormat.mps)

_, prob = LpProblem.fromMPS(mps_path)
print(str(prob))
```

With the corrent version of pyomo, we have the linear problem: 

```
Example-mix-integer-linear-problem:
MINIMIZE
3.0*x1 + 2.0*x2 + 0
SUBJECT TO
c_l_x4_: 4 x1 + 3 x2 >= 10

c_u_x5_: x1 + 2 x2 <= 7

VARIABLES
x1 Continuous
x2 Continuous
```

Note that the variable "x1" was suppose to be a integer. Now, that is the linear problem with the changes. 

```
Example-mix-integer-linear-problem:
MINIMIZE
3.0*x1 + 2.0*x2 + 0
SUBJECT TO
c_l_x4_: 4 x1 + 3 x2 >= 10

c_u_x5_: x1 + 2 x2 <= 7

VARIABLES
0 <= x1 Integer
x2 Continuous
```

## Changes proposed in this PR:
- Changing the mps format to mark integer variables correctly following the mps files configuration (https://lpsolve.sourceforge.net/5.0/mps-format.htm).

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by mrmundt** [comment]
Hi @Raphael-D-F-Gomes - Please read the [Contributing guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) to see how to resolve the linting failure.

**Review by mrmundt — CHANGES_REQUESTED** [review]
I will ask you to write a small test for this. You can do a small baseline comparison test to ensure that behavior is correct.

**Review by mrmundt — APPROVED** [review]

File: pyomo/repn/plugins/mps.py [hunk 1]
```diff
@@ -515,10 +515,22 @@ def yield_all_constraints():
...
+                if vardata.is_integer() and not set_integer:
+                    set_integer = True
+                    output_file.write(
+                        "     %s %s %s\n" % ("INT", "'MARKER'", "'INTORG'")
+                    )
+                if not vardata.is_integer() and set_integer:
+                    set_integer = False
+                    output_file.write(
+                        "     %s %s %s\n" % ("INTEND", "'MARKER'", "'INTEND'")
+                    )
```
Comment (jsiirola) [inline pyomo/repn/plugins/mps.py:532]: Is it an issue that the marker name is hard-coded? It looks like 
- Gurobi [explicitly ignores](https://www.gurobi.com/documentation/current/refman/mps_format.html) the marker names
- Cplex requires the marker name to [differ](https://www.ibm.com/docs/en/icos/22.1.0?topic=extensions-integer-variables-in-mps-files) from the preceeding / subsequent variable names (and says nothing about uniqueness, although the example hints that uniqueness might be required?).  Note that a marker name could conflict with a column name, causing issues with cplex.
- Mosek [recommends markers](https://docs.mosek.com/latest/cxxfusion/mps-format.html#integer-variables) NOT be used.  They include it for backwards compatibility and ignore the name field.
- Xpress [does not comment ](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/nonlinear/HTML/chapExtMPSformat.html) on the use of markers
- Lpsolve [requires](https://lpsolve.sourceforge.net/5.5/mps-format.htm) a name but does not make any statement as to its uniqueness

In addition, will solvers accept duplicate marker names if there are multiple groups of integer variables?

File: pyomo/repn/tests/mps/knapsack_problem_binary_variable_declaration.mps.baseline [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     x(_4_) obj 4
+     x(_4_) c_l_const1_ 35
+     x(_5_) obj 5
+     x(_5_) c_l_const1_ 29
+     x(_6_) obj 4
+     x(_6_) c_l_const1_ 8
+     x(_7_) obj 3
+     x(_7_) c_l_const1_ 31
+     x(_8_) obj 1
+     x(_8_) c_l_const1_ 18
```
Comment (jsiirola) [inline pyomo/repn/tests/mps/knapsack_problem_binary_variable_declaration.mps.baseline:27]: Several solvers documentation indicate that INTEND is required.  The logic should be updated to ensure that the INTEND marker is always emitted if `set_integer` is True at the end of the COLUMNS section.

File: pyomo/repn/plugins/mps.py [hunk 3]
```diff
@@ -515,10 +515,22 @@ def yield_all_constraints():
         column_template = "     %s %s %" + self._precision_string + "\n"
         output_file.write("COLUMNS\n")
         cnt = 0
+        set_integer = False
```
Comment (jsiirola) [inline pyomo/repn/plugins/mps.py:518]: A better name for this might be `in_integer_section`

**Review by jsiirola — DISMISSED** [review]
There are a number of changes that I think should be made.

In addition, which solver are you testing against?  Or, more specifically, which solver requires the INTORG/INTEND markers?  Looking into several solvers' documentation (Cplex, Mosek), they indicate that the markers are not necessary.

**Comment by Raphael-D-F-Gomes** [comment]
> There are a number of changes that I think should be made.
> 
> In addition, which solver are you testing against? Or, more specifically, which solver requires the INTORG/INTEND markers? Looking into several solvers' documentation (Cplex, Mosek), they indicate that the markers are not necessary.

The requested modifications have been duly assessed and subsequently implemented.

As outlined in the associated [document](https://plato.asu.edu/cplex_mps.pdf), there exist two distinct methods for specifying integer variables within the MPS format: one involving markers (INTORG, INTEND) in the COLUMNS section, and the other employing types (BV, LI, SC, UI) in the BOUNDS section. It is noteworthy that the present iteration of Pyomo exclusively supports the latter method for declaring these variables. However, it is worth mentioning that certain solvers, such as [lpsolve](https://lpsolve.sourceforge.net/5.0/mps-format.htm) and the PuLP library, acknowledge the former approach when parsing MPS files.

image


In light of enhancing the robustness of the Pyomo library, it is proposed to introduce the capability to include integer markers within the Pyomo MPS writer. Consequently, a flag has been introduced within the WriterFactory, enabling users to determine whether or not to include integer markers during the writing process.

**Review by mrmundt — APPROVED** [review]
Pending review from @blnicho or @jsiirola , I think this is good to go.

**Review by blnicho — APPROVED** [review]

**Pull request merged**